code/globalincs/pool.h

```diff
@@ -0,0 +1,404 @@
+	//Store a value in the vector
+	//TODO: Handle capped stuff
+	//This is kind of in appropriate for the actual usage this is primarily intended to replace
+	//to the point where if we want this functionality it might need to be a seperate data structure for sake of strictness
+	optional<pool_index> store(T &n){
+		if (known_empty.empty()) {
+			if (capped && storage.size()>=cap) {
+				return tl::nullopt;
+			}
+			return add_new(n);
+		}
+		return use_free(n);
+	};
+
+	void remove(optional<pool_index> i){
+		if (!i.has_value())
+			return;
+		remove(i.value());
+	};
+	void remove(pool_index i){
+		if (i.null())
+			throw PoolExceptionNullIndex("Attempted to remove from SCP_Pool with a null index");
+		if (i.i() >= storage.size())
+			throw PoolExceptionOutOfRange("Attempted to remove from SCP_Pool with out-of-range index");
+		if (storage[i.i()].generation != i.g() ) 
+			throw PoolExceptionStaleIndex("Attempted to remove from SCP_Pool with stale index");
+		reset_slot(i.i());
+	};
+	bool check(pool_index i){
+		if(i.null())
+			return false;
+		auto * t = &operator[](i);
+		return (t!=nullptr);
+	}
+
+
+
+	//iterator shit.
+	struct iter{
+		using iterator_category = std::input_iterator_tag;
+		using index = int;
+		using storage_t = vector<gEntry<T>>;
+		using storage_p = vector<gEntry<T>>*;
+
+		iter(SCP_Pool<T> *vec_ptr, storage_p storage_ptr){
+			storage = storage_ptr;
+			vec = vec_ptr;
+		};
+		iter(SCP_Pool<T> *vec_ptr, storage_p storage_ptr, int start_pos){
+			storage = storage_ptr;
+			vec = vec_ptr;
+			inner_position = start_pos;
+			if(inner_position < storage->size())
+			{
+				if(storage->at(inner_position).stored == tl::nullopt){
+					(*this)++;
+				}
+				else {
+					full_position =pool_index(inner_position,storage->at(inner_position).generation);
+				}
+			}
+		};
+		bool operator==(const iter& rhs){
+			return inner_position == rhs.inner_position;
+		};
+		bool operator<(const iter& rhs){
+			return inner_position < rhs.inner_position;
+		};
+		iter& operator++(){
+			inner_position++;
+			//we want to skip empty slots
+			while(inner_position < storage->size()
+					&& storage->at(inner_position).stored == tl::nullopt 
+				  ){
+				inner_position++;
+				//the while condition isn't short-circuiting so we're getting invalid accesses.
+				if(inner_position==storage->size()) break;
+			} 
+			if(inner_position < storage->size()){
+			   if(storage->at(inner_position).stored != tl::nullopt){
+				full_position.set(inner_position,storage->at(inner_position).generation);
+			   }
+			   else{
+				 full_position.nullify();
+			   }
+			}
+			else{full_position.nullify();}
+			return *this;
+		};
+		iter& operator++(int i){
+			operator++();
+			return *this;
+		}
+		T* operator*(){
+			//vec->get_pointer(full_position);
+			if(vec->check(full_position))
+				return &vec->operator[](full_position);
+			return nullptr;
+		};
+		T& operator&(){
+//			T t = *vec[full_position];
+			return vec->operator[](full_position);
+		};
```

Comment:
It is a _bad_ idea to overload operator& and not returning a pointer.
Really, there is no real reason in most cases, and definitely not here, to overload it.
Just let the built in operator return the address of the iterator when used.

Reply:
I've been trying to figure out where I got in my head to do this and failed so far. It was either some guidance I read somewhere about doing range-based for loop support, or just trying to minimize the change in client code.
To the latter, if I follow your guidance on return types for `*` and `->`, then in trails.cpp when I need to get a trail pointer out of an iterator, the syntax becomes `trail * trailp=iter.operator->();` or alternatively `trial *trailp = &*iter;` works. If I make `operator&()` return a pointer to the current item then `trail * = &iter;` it works just fine, at least. Not sure if that's a safe practice in this case? Certainly makes it hard to pass around pointers to an iterator, I suppose. For the working branch I'm now using the `&*` in the relevant places

Reply:
So, ``&*iter`` would in fact be the preferred way of getting a pointer to a trail.
_However_, as it is a pointer to something internally stored in a vector, it's not allowed to live longer than a single frame anyways, in which case its fine that it looks ugly, cause you really should be using references unless its for interfacing with legacy code that is built on pointers.

Reply:
Alright, unless I encounter a compelling reason to want to bring it back I'll remove this overload.